<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Get a GEO object from NCBI or file — getGEO • GEOquery</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Get a GEO object from NCBI or file — getGEO"><meta name="description" content="This function is the main user-level function in the GEOquery package.  It
directs the download (if no filename is specified) and parsing of a GEO SOFT
format file into an R data structure specifically designed to make access to
each of the important parts of the GEO SOFT format easily accessible."><meta property="og:description" content="This function is the main user-level function in the GEOquery package.  It
directs the download (if no filename is specified) and parsing of a GEO SOFT
format file into an R data structure specifically designed to make access to
each of the important parts of the GEO SOFT format easily accessible."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">GEOquery</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.73.5</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/GEOquery.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/seandavi/GEOquery/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Get a GEO object from NCBI or file</h1>
      <small class="dont-index">Source: <a href="https://github.com/seandavi/GEOquery/blob/main/R/getGEO.R" class="external-link"><code>R/getGEO.R</code></a></small>
      <div class="d-none name"><code>getGEO.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function is the main user-level function in the GEOquery package.  It
directs the download (if no filename is specified) and parsing of a GEO SOFT
format file into an R data structure specifically designed to make access to
each of the important parts of the GEO SOFT format easily accessible.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">getGEO</span><span class="op">(</span></span>
<span>  GEO <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  filename <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  destdir <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  GSElimits <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  GSEMatrix <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  AnnotGPL <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  getGPL <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  parseCharacteristics <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-geo">GEO<a class="anchor" aria-label="anchor" href="#arg-geo"></a></dt>
<dd><p>A character string representing a GEO object for download and
parsing.  (eg., 'GDS505','GSE2','GSM2','GPL96')</p></dd>


<dt id="arg-filename">filename<a class="anchor" aria-label="anchor" href="#arg-filename"></a></dt>
<dd><p>The filename of a previously downloaded GEO SOFT format file
or its gzipped representation (in which case the filename must end in .gz).
Either one of GEO or filename may be specified, not both.  GEO series matrix
files are also handled.  Note that since a single file is being parsed, the
return value is not a list of esets, but a single eset when GSE matrix files
are parsed.</p></dd>


<dt id="arg-destdir">destdir<a class="anchor" aria-label="anchor" href="#arg-destdir"></a></dt>
<dd><p>The destination directory for any downloads.  Defaults to the
architecture-dependent tempdir.  You may want to specify a different
directory if you want to save the file for later use.  Doing so is a good
idea if you have a slow connection, as some of the GEO files are HUGE!</p></dd>


<dt id="arg-gselimits">GSElimits<a class="anchor" aria-label="anchor" href="#arg-gselimits"></a></dt>
<dd><p>This argument can be used to load only a contiguous subset
of the GSMs from a GSE.  It should be specified as a vector of length 2
specifying the start and end (inclusive) GSMs to load.  This could be useful
for splitting up large GSEs into more manageable parts, for example.</p></dd>


<dt id="arg-gsematrix">GSEMatrix<a class="anchor" aria-label="anchor" href="#arg-gsematrix"></a></dt>
<dd><p>A boolean telling GEOquery whether or not to use GSE Series
Matrix files from GEO.  The parsing of these files can be many
orders-of-magnitude faster than parsing the GSE SOFT format files.  Defaults
to TRUE, meaning that the SOFT format parsing will not occur; set to FALSE
if you for some reason need other columns from the GSE records.</p></dd>


<dt id="arg-annotgpl">AnnotGPL<a class="anchor" aria-label="anchor" href="#arg-annotgpl"></a></dt>
<dd><p>A boolean defaulting to FALSE as to whether or not to use
the Annotation GPL information.  These files are nice to use because they
contain up-to-date information remapped from Entrez Gene on a regular basis.
However, they do not exist for all GPLs; in general, they are only available
for GPLs referenced by a GDS</p></dd>


<dt id="arg-getgpl">getGPL<a class="anchor" aria-label="anchor" href="#arg-getgpl"></a></dt>
<dd><p>A boolean defaulting to TRUE as to whether or not to download
and include GPL information when getting a GSEMatrix file.  You may want to
set this to FALSE if you know that you are going to annotate your
featureData using Bioconductor tools rather than relying on information
provided through NCBI GEO.  Download times can also be greatly reduced by
specifying FALSE.</p></dd>


<dt id="arg-parsecharacteristics">parseCharacteristics<a class="anchor" aria-label="anchor" href="#arg-parsecharacteristics"></a></dt>
<dd><p>A boolean defaulting to TRUE as to whether or not
to parse the characteristics information (if available) for a GSE Matrix file.
Set this to FALSE if you experience trouble while parsing the characteristics.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>An object of the appropriate class (GDS, GPL, GSM, or GSE) is
returned.  If the GSEMatrix option is used, then a list of ExpressionSet
objects is returned, one for each SeriesMatrix file associated with the GSE
accesion.  If the filename argument is used in combination with a GSEMatrix
file, then the return value is a single ExpressionSet.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>getGEO functions to download and parse information available from NCBI GEO
(<a href="http://www.ncbi.nlm.nih.gov/geo" class="external-link">http://www.ncbi.nlm.nih.gov/geo</a>).  Here are some details about what
is avaible from GEO.  All entity types are handled by getGEO and essentially
any information in the GEO SOFT format is reflected in the resulting data
structure.</p>
<p>From the GEO website:</p>
<p>The Gene Expression Omnibus (GEO) from NCBI serves as a public repository
for a wide range of high-throughput experimental data. These data include
single and dual channel microarray-based experiments measuring mRNA, genomic
DNA, and protein abundance, as well as non-array techniques such as serial
analysis of gene expression (SAGE), and mass spectrometry proteomic data. At
the most basic level of organization of GEO, there are three entity types
that may be supplied by users: Platforms, Samples, and Series.
Additionally, there is a curated entity called a GEO dataset.</p>
<p>A Platform record describes the list of elements on the array (e.g., cDNAs,
oligonucleotide probesets, ORFs, antibodies) or the list of elements that
may be detected and quantified in that experiment (e.g., SAGE tags,
peptides). Each Platform record is assigned a unique and stable GEO
accession number (GPLxxx). A Platform may reference many Samples that have
been submitted by multiple submitters.</p>
<p>A Sample record describes the conditions under which an individual Sample
was handled, the manipulations it underwent, and the abundance measurement
of each element derived from it. Each Sample record is assigned a unique and
stable GEO accession number (GSMxxx). A Sample entity must reference only
one Platform and may be included in multiple Series.</p>
<p>A Series record defines a set of related Samples considered to be part of a
group, how the Samples are related, and if and how they are ordered. A
Series provides a focal point and description of the experiment as a whole.
Series records may also contain tables describing extracted data, summary
conclusions, or analyses. Each Series record is assigned a unique and stable
GEO accession number (GSExxx).</p>
<p>GEO DataSets (GDSxxx) are curated sets of GEO Sample data. A GDS record
represents a collection of biologically and statistically comparable GEO
Samples and forms the basis of GEO's suite of data display and analysis
tools. Samples within a GDS refer to the same Platform, that is, they share
a common set of probe elements. Value measurements for each Sample within a
GDS are assumed to be calculated in an equivalent manner, that is,
considerations such as background processing and normalization are
consistent across the dataset. Information reflecting experimental design is
provided through GDS subsets.</p>
    </div>
    <div class="section level2">
    <h2 id="warning-">Warning <a class="anchor" aria-label="anchor" href="#warning-"></a></h2>
    <p>Some of the files that are downloaded, particularly those
associated with GSE entries from GEO are absolutely ENORMOUS and parsing
them can take quite some time and memory.  So, particularly when working
with large GSE entries, expect that you may need a good chunk of memory and
that coffee may be involved when parsing....</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><code><a href="getGEOfile.html">getGEOfile</a></code></p></div>
    </div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>Sean Davis</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">gds</span> <span class="op">&lt;-</span> <span class="fu">getGEO</span><span class="op">(</span><span class="st">'GDS10'</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">gds</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> An object of class "GDS"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> channel_count </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "1"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> dataset_id </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [1] "GDS10" "GDS10" "GDS10" "GDS10" "GDS10" "GDS10" "GDS10" "GDS10" "GDS10"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [10] "GDS10" "GDS10" "GDS10"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> description </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [1] "Examination of spleen and thymus of type 1 diabetes nonobese diabetic (NOD) mouse, four NOD-derived diabetes-resistant congenic strains and two nondiabetic control strains."</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [2] "spleen"                                                                                                                                                                      </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [3] "thymus"                                                                                                                                                                      </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [4] "NOD"                                                                                                                                                                         </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [5] "Idd3"                                                                                                                                                                        </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [6] "Idd5"                                                                                                                                                                        </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [7] "Idd3+Idd5"                                                                                                                                                                   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [8] "Idd9"                                                                                                                                                                        </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [9] "B10.H2g7"                                                                                                                                                                    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [10] "B10.H2g7 Idd3"                                                                                                                                                               </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [11] "diabetic"                                                                                                                                                                    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [12] "diabetic-resistant"                                                                                                                                                          </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [13] "nondiabetic"                                                                                                                                                                 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> email </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "geo@ncbi.nlm.nih.gov"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> feature_count </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "39114"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> institute </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "NCBI NLM NIH"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> name </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Gene Expression Omnibus (GEO)"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> order </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "none"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> platform </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "GPL24"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> platform_organism </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Mus musculus"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> platform_technology_type </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "in situ oligonucleotide"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> pubmed_id </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "11827943"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ref </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Nucleic Acids Res. 2005 Jan 1;33 Database Issue:D562-6"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> reference_series </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "GSE11"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> sample_count </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "28"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> sample_id </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [1] "GSM582,GSM583,GSM584,GSM585,GSM586,GSM587,GSM588,GSM589,GSM590,GSM591,GSM592,GSM593,GSM594,GSM595"              </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [2] "GSM596,GSM597,GSM598,GSM599,GSM600,GSM601,GSM602,GSM603,GSM604,GSM605,GSM606,GSM607,GSM608,GSM609"              </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [3] "GSM582,GSM589,GSM596,GSM603"                                                                                    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [4] "GSM583,GSM590,GSM597,GSM604"                                                                                    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [5] "GSM584,GSM591,GSM598,GSM605"                                                                                    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [6] "GSM585,GSM592,GSM599,GSM606"                                                                                    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [7] "GSM586,GSM593,GSM600,GSM607"                                                                                    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [8] "GSM587,GSM594,GSM601,GSM608"                                                                                    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [9] "GSM588,GSM595,GSM602,GSM609"                                                                                    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [10] "GSM582,GSM589,GSM596,GSM603"                                                                                    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [11] "GSM583,GSM590,GSM597,GSM604,GSM584,GSM591,GSM598,GSM605,GSM585,GSM592,GSM599,GSM606,GSM586,GSM593,GSM600,GSM607"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [12] "GSM587,GSM594,GSM601,GSM608,GSM588,GSM595,GSM602,GSM609"                                                        </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> sample_organism </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Mus musculus"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> sample_type </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "RNA"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> title </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Type 1 diabetes gene expression profiling"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> type </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [1] "Expression profiling by array" "tissue"                       </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [3] "tissue"                        "strain"                       </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [5] "strain"                        "strain"                       </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [7] "strain"                        "strain"                       </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [9] "strain"                        "strain"                       </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [11] "disease state"                 "disease state"                </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [13] "disease state"                </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> update_date </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Jul 15 2003"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> value_type </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "count"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> web_link </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "http://www.ncbi.nlm.nih.gov/geo"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> An object of class "GEODataTable"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ****** Column Descriptions ******</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    sample tissue        strain      disease.state</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1  GSM582 spleen           NOD           diabetic</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2  GSM589 spleen           NOD           diabetic</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3  GSM583 spleen          Idd3 diabetic-resistant</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4  GSM590 spleen          Idd3 diabetic-resistant</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5  GSM584 spleen          Idd5 diabetic-resistant</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6  GSM591 spleen          Idd5 diabetic-resistant</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 7  GSM585 spleen     Idd3+Idd5 diabetic-resistant</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 8  GSM592 spleen     Idd3+Idd5 diabetic-resistant</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 9  GSM586 spleen          Idd9 diabetic-resistant</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 10 GSM593 spleen          Idd9 diabetic-resistant</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 11 GSM587 spleen      B10.H2g7        nondiabetic</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 12 GSM594 spleen      B10.H2g7        nondiabetic</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 13 GSM588 spleen B10.H2g7 Idd3        nondiabetic</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 14 GSM595 spleen B10.H2g7 Idd3        nondiabetic</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 15 GSM596 thymus           NOD           diabetic</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 16 GSM603 thymus           NOD           diabetic</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 17 GSM597 thymus          Idd3 diabetic-resistant</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 18 GSM604 thymus          Idd3 diabetic-resistant</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 19 GSM598 thymus          Idd5 diabetic-resistant</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 20 GSM605 thymus          Idd5 diabetic-resistant</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 21 GSM599 thymus     Idd3+Idd5 diabetic-resistant</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 22 GSM606 thymus     Idd3+Idd5 diabetic-resistant</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 23 GSM600 thymus          Idd9 diabetic-resistant</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 24 GSM607 thymus          Idd9 diabetic-resistant</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 25 GSM601 thymus      B10.H2g7        nondiabetic</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 26 GSM608 thymus      B10.H2g7        nondiabetic</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 27 GSM602 thymus B10.H2g7 Idd3        nondiabetic</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 28 GSM609 thymus B10.H2g7 Idd3        nondiabetic</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                                        description</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1            Value for GSM582: NOD_S1; src: Spleen</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2            Value for GSM589: NOD_S2; src: Spleen</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3           Value for GSM583: Idd3_S1; src: Spleen</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4           Value for GSM590: Idd3_S2; src: Spleen</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5           Value for GSM584: Idd5_S1; src: Spleen</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6           Value for GSM591: Idd5_S2; src: Spleen</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 7         Value for GSM585: Idd3+5_S1; src: Spleen</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 8         Value for GSM592: Idd3+5_S2; src: Spleen</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 9           Value for GSM586: Idd9_S1; src: Spleen</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 10          Value for GSM593: Idd9_S2; src: Spleen</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 11      Value for GSM587: B10.H2g7_S1; src: Spleen</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 12      Value for GSM594: B10.H2g7_S2; src: Spleen</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 13 Value for GSM588: B10.H2g7 Idd3_S1; src: Spleen</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 14 Value for GSM595: B10.H2g7 Idd3_S2; src: Spleen</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 15           Value for GSM596: NOD_T1; src: Thymus</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 16           Value for GSM603: NOD_T2; src: Thymus</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 17          Value for GSM597: Idd3_T1; src: Thymus</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 18          Value for GSM604: Idd3_T2; src: Thymus</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 19          Value for GSM598: Idd5_T1; src: Thymus</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 20          Value for GSM605: Idd5_T2; src: Thymus</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 21        Value for GSM599: Idd3+5_T1; src: Thymus</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 22        Value for GSM606: Idd3+5_T2; src: Thymus</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 23          Value for GSM600: Idd9_T1; src: Thymus</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 24          Value for GSM607: Idd9_T2; src: Thymus</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 25      Value for GSM601: B10.H2g7_T1; src: Thymus</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 26      Value for GSM608: B10.H2g7_T2; src: Thymus</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 27 Value for GSM602: B10.H2g7 Idd3_T1; src: Thymus</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 28 Value for GSM609: B10.H2g7 Idd3_T2; src: Thymus</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ****** Data Table ******</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">gse</span> <span class="op">&lt;-</span> <span class="fu">getGEO</span><span class="op">(</span><span class="st">'GSE10'</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Found 1 file(s)</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> GSE10_series_matrix.txt.gz</span>
<span class="r-in"><span><span class="co"># Returns a list, so look at first item</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">gse</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ExpressionSet (storageMode: lockedEnvironment)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> assayData: 96903 features, 4 samples </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   element names: exprs </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> protocolData: none</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> phenoData</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   sampleNames: GSM571 GSM572 GSM573 GSM574</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   varLabels: title geo_accession ... data_row_count (34 total)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   varMetadata: labelDescription</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> featureData</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   featureNames: AAAAAAAAAA AAAAAAAAAC ... TTTTTTTTTT (96903 total)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   fvarLabels: TAG GI</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   fvarMetadata: Column Description labelDescription</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> experimentData: use 'experimentData(object)'</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   pubMedIds: 11756676 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Annotation: GPL4 </span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Sean Davis.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

