<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Using the GEOquery Package • GEOquery</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Using the GEOquery Package">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">GEOquery</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.73.4</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item"><a class="nav-link" href="../articles/GEOquery.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/seandavi/GEOquery/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Using the GEOquery Package</h1>
                        <h4 data-toc-skip class="author">Sean Davis</h4>
            
            <h4 data-toc-skip class="date">September 21, 2014</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/seandavi/GEOquery/blob/main/vignettes/GEOquery.Rmd" class="external-link"><code>vignettes/GEOquery.Rmd</code></a></small>
      <div class="d-none name"><code>GEOquery.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="overview-of-geo">Overview of GEO<a class="anchor" aria-label="anchor" href="#overview-of-geo"></a>
</h2>
<p>The NCBI Gene Expression Omnibus (GEO) serves as a public repository
for a wide range of high-throughput experimental data. These data
include single and dual channel microarray-based experiments measuring
mRNA, genomic DNA, and protein abundance, as well as non-array
techniques such as serial analysis of gene expression (SAGE), mass
spectrometry proteomic data, and high-throughput sequencing data.</p>
<p>At the most basic level of organization of GEO, there are four basic
entity types. The first three (Sample, Platform, and Series) are
supplied by users; the fourth, the dataset, is compiled and curated by
GEO staff from the user-submitted data. See <a href="https://www.ncbi.nlm.nih.gov/geo/" class="external-link">the GEO home page</a> for more
information.</p>
<div class="section level3">
<h3 id="platforms">Platforms<a class="anchor" aria-label="anchor" href="#platforms"></a>
</h3>
<p>A Platform record describes the list of elements on the array (e.g.,
cDNAs, oligonucleotide probesets, ORFs, antibodies) or the list of
elements that may be detected and quantified in that experiment (e.g.,
SAGE tags, peptides). Each Platform record is assigned a unique and
stable GEO accession number (GPLxxx). A Platform may reference many
Samples that have been submitted by multiple submitters.</p>
</div>
<div class="section level3">
<h3 id="samples">Samples<a class="anchor" aria-label="anchor" href="#samples"></a>
</h3>
<p>A Sample record describes the conditions under which an individual
Sample was handled, the manipulations it underwent, and the abundance
measurement of each element derived from it. Each Sample record is
assigned a unique and stable GEO accession number (GSMxxx). A Sample
entity must reference only one Platform and may be included in multiple
Series.</p>
</div>
<div class="section level3">
<h3 id="series">Series<a class="anchor" aria-label="anchor" href="#series"></a>
</h3>
<p>A Series record defines a set of related Samples considered to be
part of a group, how the Samples are related, and if and how they are
ordered. A Series provides a focal point and description of the
experiment as a whole. Series records may also contain tables describing
extracted data, summary conclusions, or analyses. Each Series record is
assigned a unique and stable GEO accession number (GSExxx). Series
records are available in a couple of formats which are handled by
GEOquery independently. The smaller and new GSEMatrix files are quite
fast to parse; a simple flag is used by GEOquery to choose to use
GSEMatrix files (see below).</p>
</div>
<div class="section level3">
<h3 id="datasets">Datasets<a class="anchor" aria-label="anchor" href="#datasets"></a>
</h3>
<p>GEO DataSets (GDSxxx) are curated sets of GEO Sample data. A GDS
record represents a collection of biologically and statistically
comparable GEO Samples and forms the basis of GEO’s suite of data
display and analysis tools. Samples within a GDS refer to the same
Platform, that is, they share a common set of probe elements. Value
measurements for each Sample within a GDS are assumed to be calculated
in an equivalent manner, that is, considerations such as background
processing and normalization are consistent across the dataset.
Information reflecting experimental design is provided through GDS
subsets.</p>
</div>
</div>
<div class="section level2">
<h2 id="getting-started-using-geoquery">Getting Started using GEOquery<a class="anchor" aria-label="anchor" href="#getting-started-using-geoquery"></a>
</h2>
<p>Getting data from GEO is really quite easy. There is only one command
that is needed, <code>getGEO</code>. This one function interprets its
input to determine how to get the data from GEO and then parse the data
into useful R data structures. Usage is quite simple. This loads the
GEOquery library.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/seandavi/GEOquery" class="external-link">GEOquery</a></span><span class="op">)</span></span></code></pre></div>
<p>Now, we are free to access any GEO accession. <em>Note that in the
following, I use a file packaged with the GEOquery package. In general,
you will use only the GEO accession, as noted in the code
comments.</em></p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># If you have network access, the more typical way to do this</span></span>
<span><span class="co"># would be to use this:</span></span>
<span><span class="co"># gds &lt;- getGEO("GDS507")</span></span>
<span><span class="va">gds</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getGEO.html">getGEO</a></span><span class="op">(</span>filename<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/GDS507.soft.gz"</span>,package<span class="op">=</span><span class="st">"GEOquery"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Now, <code>gds</code> contains the R data structure (of class
<code>GDS</code>) that represents the GDS507 entry from GEO. You’ll note
that the filename used to store the download was output to the screen
(but not saved anywhere) for later use to a call to
<code>getGEO(filename=...)</code>.</p>
<p>We can do the same with any other GEO accession, such as
<code>GSM11805</code>, a GEO sample.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># If you have network access, the more typical way to do this</span></span>
<span><span class="co"># would be to use this:</span></span>
<span><span class="co"># gds &lt;- getGEO("GSM11805")</span></span>
<span><span class="va">gsm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getGEO.html">getGEO</a></span><span class="op">(</span>filename<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/GSM11805.txt.gz"</span>,package<span class="op">=</span><span class="st">"GEOquery"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="geoquery-data-structures">GEOquery Data Structures<a class="anchor" aria-label="anchor" href="#geoquery-data-structures"></a>
</h2>
<p>The GEOquery data structures really come in two forms. The first,
comprising <code>GDS</code>, <code>GPL</code>, and <code>GSM</code> all
behave similarly and accessors have similar effects on each. The fourth
GEOquery data structure, <code>GSE</code> is a composite data type made
up of a combination of <code>GSM</code> and <code>GPL</code> objects. I
will explain the first three together first.</p>
<div class="section level3">
<h3 id="the-gds-gsm-and-gpl-classes">The GDS, GSM, and GPL classes<a class="anchor" aria-label="anchor" href="#the-gds-gsm-and-gpl-classes"></a>
</h3>
<p>Each of these classes is comprised of a metadata header (taken nearly
verbatim from the SOFT format header) and a GEODataTable. The
GEODataTable has two simple parts, a Columns part which describes the
column headers on the Table part. There is also a <code>show</code>
method for each class. For example, using the gsm from above:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Look at gsm metadata:</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="../reference/GEOData-accessors.html">Meta</a></span><span class="op">(</span><span class="va">gsm</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## $channel_count</span></span>
<span><span class="co">## [1] "1"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $comment</span></span>
<span><span class="co">## [1] "Raw data provided as supplementary file"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $contact_address</span></span>
<span><span class="co">## [1] "715 Albany Street, E613B"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $contact_city</span></span>
<span><span class="co">## [1] "Boston"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $contact_country</span></span>
<span><span class="co">## [1] "USA"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $contact_department</span></span>
<span><span class="co">## [1] "Genetics and Genomics"</span></span></code></pre>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Look at data associated with the GSM:</span></span>
<span><span class="co"># but restrict to only first 5 rows, for brevity</span></span>
<span><span class="fu"><a href="../reference/GEOData-accessors.html">Table</a></span><span class="op">(</span><span class="va">gsm</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>,<span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">##           ID_REF  VALUE ABS_CALL</span></span>
<span><span class="co">## 1 AFFX-BioB-5_at  953.9        P</span></span>
<span><span class="co">## 2 AFFX-BioB-M_at 2982.8        P</span></span>
<span><span class="co">## 3 AFFX-BioB-3_at 1657.9        P</span></span>
<span><span class="co">## 4 AFFX-BioC-5_at 2652.7        P</span></span>
<span><span class="co">## 5 AFFX-BioC-3_at 2019.5        P</span></span></code></pre>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Look at Column descriptions:</span></span>
<span><span class="fu"><a href="../reference/GEOData-accessors.html">Columns</a></span><span class="op">(</span><span class="va">gsm</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##     Column</span></span>
<span><span class="co">## 1         </span></span>
<span><span class="co">## 2    VALUE</span></span>
<span><span class="co">## 3 ABS_CALL</span></span>
<span><span class="co">##                                                                  Description</span></span>
<span><span class="co">## 1                                                                   ID_REF =</span></span>
<span><span class="co">## 2                         MAS 5.0 Statistical Algorithm (mean scaled to 500)</span></span>
<span><span class="co">## 3 MAS 5.0 Absent, Marginal, Present call  with Alpha1 = 0.05, Alpha2 = 0.065</span></span></code></pre>
<p>The <code>GPL</code> class behaves exactly as the <code>GSM</code>
class. However, the <code>GDS</code> class has a bit more information
associated with the <code>Columns</code> method:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/GEOData-accessors.html">Columns</a></span><span class="op">(</span><span class="va">gds</span><span class="op">)</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">##      sample disease.state individual</span></span>
<span><span class="co">## 1  GSM11815           RCC        035</span></span>
<span><span class="co">## 2  GSM11832           RCC        023</span></span>
<span><span class="co">## 3  GSM12069           RCC        001</span></span>
<span><span class="co">## 4  GSM12083           RCC        005</span></span>
<span><span class="co">## 5  GSM12101           RCC        011</span></span>
<span><span class="co">## 6  GSM12106           RCC        032</span></span>
<span><span class="co">## 7  GSM12274           RCC          2</span></span>
<span><span class="co">## 8  GSM12299           RCC          3</span></span>
<span><span class="co">## 9  GSM12412           RCC          4</span></span>
<span><span class="co">## 10 GSM11810        normal        035</span></span>
<span><span class="co">## 11 GSM11827        normal        023</span></span>
<span><span class="co">## 12 GSM12078        normal        001</span></span>
<span><span class="co">## 13 GSM12099        normal        005</span></span>
<span><span class="co">## 14 GSM12269        normal          1</span></span>
<span><span class="co">## 15 GSM12287        normal          2</span></span>
<span><span class="co">## 16 GSM12301        normal          3</span></span>
<span><span class="co">## 17 GSM12448        normal          4</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="the-gse-class">The GSE class<a class="anchor" aria-label="anchor" href="#the-gse-class"></a>
</h3>
<p>The <code>GSE</code> entity is the most confusing of the GEO
entities. A GSE entry can represent an arbitrary number of samples run
on an arbitrary number of platforms. The <code>GSE</code> class has a
metadata section, just like the other classes. However, it doesn’t have
a GEODataTable. Instead, it contains two lists, accessible using the
<code>GPLList</code> and <code>GSMList</code> methods, that are each
lists of <code>GPL</code> and <code>GSM</code> objects. To show an
example:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Again, with good network access, one would do:</span></span>
<span><span class="co"># gse &lt;- getGEO("GSE781",GSEMatrix=FALSE)</span></span>
<span><span class="va">gse</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getGEO.html">getGEO</a></span><span class="op">(</span>filename<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/GSE781_family.soft.gz"</span>,package<span class="op">=</span><span class="st">"GEOquery"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="../reference/GEOData-accessors.html">Meta</a></span><span class="op">(</span><span class="va">gse</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## $contact_address</span></span>
<span><span class="co">## [1] "715 Albany Street, E613B"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $contact_city</span></span>
<span><span class="co">## [1] "Boston"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $contact_country</span></span>
<span><span class="co">## [1] "USA"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $contact_department</span></span>
<span><span class="co">## [1] "Genetics and Genomics"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $contact_email</span></span>
<span><span class="co">## [1] "mlenburg@bu.edu"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $contact_fax</span></span>
<span><span class="co">## [1] "617-414-1646"</span></span></code></pre>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># names of all the GSM objects contained in the GSE</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu"><a href="../reference/GEOData-accessors.html">GSMList</a></span><span class="op">(</span><span class="va">gse</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] "GSM11805" "GSM11810" "GSM11814" "GSM11815" "GSM11823" "GSM11827"</span></span>
<span><span class="co">##  [7] "GSM11830" "GSM11832" "GSM12067" "GSM12069" "GSM12075" "GSM12078"</span></span>
<span><span class="co">## [13] "GSM12079" "GSM12083" "GSM12098" "GSM12099" "GSM12100" "GSM12101"</span></span>
<span><span class="co">## [19] "GSM12105" "GSM12106" "GSM12268" "GSM12269" "GSM12270" "GSM12274"</span></span>
<span><span class="co">## [25] "GSM12283" "GSM12287" "GSM12298" "GSM12299" "GSM12300" "GSM12301"</span></span>
<span><span class="co">## [31] "GSM12399" "GSM12412" "GSM12444" "GSM12448"</span></span></code></pre>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># and get the first GSM object on the list</span></span>
<span><span class="fu"><a href="../reference/GEOData-accessors.html">GSMList</a></span><span class="op">(</span><span class="va">gse</span><span class="op">)</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">## An object of class "GSM"</span></span>
<span><span class="co">## channel_count </span></span>
<span><span class="co">## [1] "1"</span></span>
<span><span class="co">## comment </span></span>
<span><span class="co">## [1] "Raw data provided as supplementary file"</span></span>
<span><span class="co">## contact_address </span></span>
<span><span class="co">## [1] "715 Albany Street, E613B"</span></span>
<span><span class="co">## contact_city </span></span>
<span><span class="co">## [1] "Boston"</span></span>
<span><span class="co">## contact_country </span></span>
<span><span class="co">## [1] "USA"</span></span>
<span><span class="co">## contact_department </span></span>
<span><span class="co">## [1] "Genetics and Genomics"</span></span>
<span><span class="co">## contact_email </span></span>
<span><span class="co">## [1] "mlenburg@bu.edu"</span></span>
<span><span class="co">## contact_fax </span></span>
<span><span class="co">## [1] "617-414-1646"</span></span>
<span><span class="co">## contact_institute </span></span>
<span><span class="co">## [1] "Boston University School of Medicine"</span></span>
<span><span class="co">## contact_name </span></span>
<span><span class="co">## [1] "Marc,E.,Lenburg"</span></span>
<span><span class="co">## contact_phone </span></span>
<span><span class="co">## [1] "617-414-1375"</span></span>
<span><span class="co">## contact_state </span></span>
<span><span class="co">## [1] "MA"</span></span>
<span><span class="co">## contact_web_link </span></span>
<span><span class="co">## [1] "http://gg.bu.edu"</span></span>
<span><span class="co">## contact_zip/postal_code </span></span>
<span><span class="co">## [1] "02130"</span></span>
<span><span class="co">## data_row_count </span></span>
<span><span class="co">## [1] "22283"</span></span>
<span><span class="co">## description </span></span>
<span><span class="co">## [1] "Age = 70; Gender = Female; Right Kidney; Adjacent Tumor Type = clear cell; Adjacent Tumor Fuhrman Grade = 3; Adjacent Tumor Capsule Penetration = true; Adjacent Tumor Perinephric Fat Invasion = true; Adjacent Tumor Renal Sinus Invasion = false; Adjacent Tumor Renal Vein Invasion = true; Scaling Target = 500; Scaling Factor = 7.09; Raw Q = 2.39; Noise = 2.60; Background = 55.24."</span></span>
<span><span class="co">## [2] "Keywords = kidney"                                                                                                                                                                                                                                                                                                                                                                           </span></span>
<span><span class="co">## [3] "Keywords = renal"                                                                                                                                                                                                                                                                                                                                                                            </span></span>
<span><span class="co">## [4] "Keywords = RCC"                                                                                                                                                                                                                                                                                                                                                                              </span></span>
<span><span class="co">## [5] "Keywords = carcinoma"                                                                                                                                                                                                                                                                                                                                                                        </span></span>
<span><span class="co">## [6] "Keywords = cancer"                                                                                                                                                                                                                                                                                                                                                                           </span></span>
<span><span class="co">## [7] "Lot batch = 2004638"                                                                                                                                                                                                                                                                                                                                                                         </span></span>
<span><span class="co">## geo_accession </span></span>
<span><span class="co">## [1] "GSM11805"</span></span>
<span><span class="co">## last_update_date </span></span>
<span><span class="co">## [1] "May 28 2005"</span></span>
<span><span class="co">## molecule_ch1 </span></span>
<span><span class="co">## [1] "total RNA"</span></span>
<span><span class="co">## organism_ch1 </span></span>
<span><span class="co">## [1] "Homo sapiens"</span></span>
<span><span class="co">## platform_id </span></span>
<span><span class="co">## [1] "GPL96"</span></span>
<span><span class="co">## series_id </span></span>
<span><span class="co">## [1] "GSE781"</span></span>
<span><span class="co">## source_name_ch1 </span></span>
<span><span class="co">## [1] "Trizol isolation of total RNA from normal tissue adjacent to Renal Cell Carcinoma"</span></span>
<span><span class="co">## status </span></span>
<span><span class="co">## [1] "Public on Nov 25 2003"</span></span>
<span><span class="co">## submission_date </span></span>
<span><span class="co">## [1] "Oct 20 2003"</span></span>
<span><span class="co">## supplementary_file </span></span>
<span><span class="co">## [1] "ftp://ftp.ncbi.nih.gov/pub/geo/DATA/supplementary/samples/GSM11nnn/GSM11805/GSM11805.CEL.gz"</span></span>
<span><span class="co">## title </span></span>
<span><span class="co">## [1] "N035 Normal Human Kidney U133A"</span></span>
<span><span class="co">## type </span></span>
<span><span class="co">## [1] "RNA"</span></span>
<span><span class="co">## An object of class "GEODataTable"</span></span>
<span><span class="co">## ****** Column Descriptions ******</span></span>
<span><span class="co">##     Column</span></span>
<span><span class="co">## 1         </span></span>
<span><span class="co">## 2    VALUE</span></span>
<span><span class="co">## 3 ABS_CALL</span></span>
<span><span class="co">##                                                                  Description</span></span>
<span><span class="co">## 1                                                                   ID_REF =</span></span>
<span><span class="co">## 2                         MAS 5.0 Statistical Algorithm (mean scaled to 500)</span></span>
<span><span class="co">## 3 MAS 5.0 Absent, Marginal, Present call  with Alpha1 = 0.05, Alpha2 = 0.065</span></span>
<span><span class="co">## ****** Data Table ******</span></span></code></pre>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># and the names of the GPLs represented</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu"><a href="../reference/GEOData-accessors.html">GPLList</a></span><span class="op">(</span><span class="va">gse</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "GPL96" "GPL97"</span></span></code></pre>
<p><em>See below for an additional, preferred method of obtaining GSE
information.</em></p>
</div>
</div>
<div class="section level2">
<h2 id="converting-to-bioconductor-expressionsets-and-limma-malists">Converting to BioConductor ExpressionSets and limma MALists<a class="anchor" aria-label="anchor" href="#converting-to-bioconductor-expressionsets-and-limma-malists"></a>
</h2>
<p>GEO datasets are (unlike some of the other GEO entities), quite
similar to the <code>limma</code> data structure <code>MAList</code> and
to the <code>Biobase</code> data structure <code>ExpressionSet</code>.
Therefore, there are two functions, <code>GDS2MA</code> and
<code>GDS2eSet</code> that accomplish that task.</p>
<div class="section level3">
<h3 id="getting-gse-series-matrix-files-as-an-expressionset">Getting GSE Series Matrix files as an ExpressionSet<a class="anchor" aria-label="anchor" href="#getting-gse-series-matrix-files-as-an-expressionset"></a>
</h3>
<p>GEO Series are collections of related experiments. In addition to
being available as SOFT format files, which are quite large, NCBI GEO
has prepared a simpler format file based on tab-delimited text. The
<code>getGEO</code> function can handle this format and will parse very
large GSEs quite quickly. The data structure returned from this parsing
is a list of ExpressionSets. As an example, we download and parse
GSE2553.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Note that GSEMatrix=TRUE is the default</span></span>
<span><span class="va">gse2553</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getGEO.html">getGEO</a></span><span class="op">(</span><span class="st">'GSE2553'</span>,GSEMatrix<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/methods/show.html" class="external-link">show</a></span><span class="op">(</span><span class="va">gse2553</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## $GSE2553_series_matrix.txt.gz</span></span>
<span><span class="co">## ExpressionSet (storageMode: lockedEnvironment)</span></span>
<span><span class="co">## assayData: 12600 features, 181 samples </span></span>
<span><span class="co">##   element names: exprs </span></span>
<span><span class="co">## protocolData: none</span></span>
<span><span class="co">## phenoData</span></span>
<span><span class="co">##   sampleNames: GSM48681 GSM48682 ... GSM48861 (181 total)</span></span>
<span><span class="co">##   varLabels: title geo_accession ... data_row_count (30 total)</span></span>
<span><span class="co">##   varMetadata: labelDescription</span></span>
<span><span class="co">## featureData</span></span>
<span><span class="co">##   featureNames: 1 2 ... 12600 (12600 total)</span></span>
<span><span class="co">##   fvarLabels: ID PenAt ... Chimeric_Cluster_IDs (13 total)</span></span>
<span><span class="co">##   fvarMetadata: Column Description labelDescription</span></span>
<span><span class="co">## experimentData: use 'experimentData(object)'</span></span>
<span><span class="co">##   pubMedIds: 16230383 </span></span>
<span><span class="co">## Annotation: GPL1977</span></span></code></pre>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/methods/show.html" class="external-link">show</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/phenoData.html" class="external-link">pData</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/phenoData.html" class="external-link">phenoData</a></span><span class="op">(</span><span class="va">gse2553</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">6</span>,<span class="fl">8</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                                                                  title type</span></span>
<span><span class="co">## GSM48681                      Patient sample ST18, Dermatofibrosarcoma  RNA</span></span>
<span><span class="co">## GSM48682                           Patient sample ST410, Ewing Sarcoma  RNA</span></span>
<span><span class="co">## GSM48683                            Patient sample ST130, Sarcoma, NOS  RNA</span></span>
<span><span class="co">## GSM48684 Patient sample ST293, Malignant Peripheral Nerve Sheath Tumor  RNA</span></span>
<span><span class="co">## GSM48685                             Patient sample ST367, Liposarcoma  RNA</span></span>
<span><span class="co">##                                  source_name_ch1</span></span>
<span><span class="co">## GSM48681                     Dermatofibrosarcoma</span></span>
<span><span class="co">## GSM48682                           Ewing Sarcoma</span></span>
<span><span class="co">## GSM48683                            Sarcoma, NOS</span></span>
<span><span class="co">## GSM48684 Malignant Peripheral Nerve Sheath Tumor</span></span>
<span><span class="co">## GSM48685                             Liposarcoma</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="converting-gds-to-an-expressionset">Converting GDS to an ExpressionSet<a class="anchor" aria-label="anchor" href="#converting-gds-to-an-expressionset"></a>
</h3>
<p>Taking our <code>gds</code> object from above, we can simply do:</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">eset</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/coercion.html">GDS2eSet</a></span><span class="op">(</span><span class="va">gds</span>,do.log2<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>Now, <code>eset</code> is an <code>ExpressionSet</code> that contains
the same information as in the GEO dataset, including the sample
information, which we can see here:</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">eset</span></span></code></pre></div>
<pre><code><span><span class="co">## ExpressionSet (storageMode: lockedEnvironment)</span></span>
<span><span class="co">## assayData: 22645 features, 17 samples </span></span>
<span><span class="co">##   element names: exprs </span></span>
<span><span class="co">## protocolData: none</span></span>
<span><span class="co">## phenoData</span></span>
<span><span class="co">##   sampleNames: GSM11815 GSM11832 ... GSM12448 (17 total)</span></span>
<span><span class="co">##   varLabels: sample disease.state individual description</span></span>
<span><span class="co">##   varMetadata: labelDescription</span></span>
<span><span class="co">## featureData</span></span>
<span><span class="co">##   featureNames: 200000_s_at 200001_at ... AFFX-TrpnX-M_at (22645 total)</span></span>
<span><span class="co">##   fvarLabels: ID Gene title ... GO:Component ID (21 total)</span></span>
<span><span class="co">##   fvarMetadata: Column labelDescription</span></span>
<span><span class="co">## experimentData: use 'experimentData(object)'</span></span>
<span><span class="co">##   pubMedIds: 14641932 </span></span>
<span><span class="co">## Annotation:</span></span></code></pre>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/phenoData.html" class="external-link">pData</a></span><span class="op">(</span><span class="va">eset</span><span class="op">)</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">##            sample disease.state individual</span></span>
<span><span class="co">## GSM11815 GSM11815           RCC        035</span></span>
<span><span class="co">## GSM11832 GSM11832           RCC        023</span></span>
<span><span class="co">## GSM12069 GSM12069           RCC        001</span></span>
<span><span class="co">## GSM12083 GSM12083           RCC        005</span></span>
<span><span class="co">## GSM12101 GSM12101           RCC        011</span></span>
<span><span class="co">## GSM12106 GSM12106           RCC        032</span></span>
<span><span class="co">## GSM12274 GSM12274           RCC          2</span></span>
<span><span class="co">## GSM12299 GSM12299           RCC          3</span></span>
<span><span class="co">## GSM12412 GSM12412           RCC          4</span></span>
<span><span class="co">## GSM11810 GSM11810        normal        035</span></span>
<span><span class="co">## GSM11827 GSM11827        normal        023</span></span>
<span><span class="co">## GSM12078 GSM12078        normal        001</span></span>
<span><span class="co">## GSM12099 GSM12099        normal        005</span></span>
<span><span class="co">## GSM12269 GSM12269        normal          1</span></span>
<span><span class="co">## GSM12287 GSM12287        normal          2</span></span>
<span><span class="co">## GSM12301 GSM12301        normal          3</span></span>
<span><span class="co">## GSM12448 GSM12448        normal          4</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="converting-gds-to-an-malist">Converting GDS to an MAList<a class="anchor" aria-label="anchor" href="#converting-gds-to-an-malist"></a>
</h3>
<p>No annotation information (called platform information by GEO) was
retrieved from because <code>ExpressionSet</code> does not contain slots
for gene information, typically. However, it is easy to obtain this
information. First, we need to know what platform this GDS used. Then,
another call to <code>getGEO</code> will get us what we need.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#get the platform from the GDS metadata</span></span>
<span><span class="fu"><a href="../reference/GEOData-accessors.html">Meta</a></span><span class="op">(</span><span class="va">gds</span><span class="op">)</span><span class="op">$</span><span class="va">platform</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "GPL97"</span></span></code></pre>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#So use this information in a call to getGEO</span></span>
<span><span class="va">gpl</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getGEO.html">getGEO</a></span><span class="op">(</span>filename<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/GPL97.annot.gz"</span>,package<span class="op">=</span><span class="st">"GEOquery"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>So, <code>gpl</code> now contains the information for GPL5 from GEO.
Unlike <code>ExpressionSet</code>, the limma <code>MAList</code> does
store gene annotation information, so we can use our newly created
<code>gpl</code> of class <code>GPL</code> in a call to
<code>GDS2MA</code> like so:</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">MA</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/coercion.html">GDS2MA</a></span><span class="op">(</span><span class="va">gds</span>,GPL<span class="op">=</span><span class="va">gpl</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">MA</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "MAList"</span></span>
<span><span class="co">## attr(,"package")</span></span>
<span><span class="co">## [1] "limma"</span></span></code></pre>
<p>Now, <code>MA</code> is of class <code>MAList</code> and contains not
only the data, but the sample information and gene information
associated with GDS507.</p>
</div>
<div class="section level3">
<h3 id="converting-gse-to-an-expressionset">Converting GSE to an ExpressionSet<a class="anchor" aria-label="anchor" href="#converting-gse-to-an-expressionset"></a>
</h3>
<p><em>First, make sure that using the method described above in the
section ``Getting GSE Series Matrix files as an ExpressionSet’’ for
using GSE Series Matrix files is not sufficient for the task, as it is
much faster and simpler.</em> If it is not (i.e., other columns from
each GSM are needed), then this method will be needed.</p>
<p>Converting a <code>GSE</code> object to an <code>ExpressionSet</code>
object currently takes a bit of R data manipulation due to the varied
data that can be stored in a <code>GSE</code> and the underlying
<code>GSM</code> and <code>GPL</code> objects. However, using a simple
example will hopefully be illustrative of the technique.</p>
<p>First, we need to make sure that all of the <code>GSMs</code> are
from the same platform:</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gsmplatforms</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="fu"><a href="../reference/GEOData-accessors.html">GSMList</a></span><span class="op">(</span><span class="va">gse</span><span class="op">)</span>,<span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span><span class="fu"><a href="../reference/GEOData-accessors.html">Meta</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">$</span><span class="va">platform_id</span><span class="op">}</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">gsmplatforms</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## $GSM11805</span></span>
<span><span class="co">## [1] "GPL96"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $GSM11810</span></span>
<span><span class="co">## [1] "GPL97"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $GSM11814</span></span>
<span><span class="co">## [1] "GPL96"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $GSM11815</span></span>
<span><span class="co">## [1] "GPL97"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $GSM11823</span></span>
<span><span class="co">## [1] "GPL96"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $GSM11827</span></span>
<span><span class="co">## [1] "GPL97"</span></span></code></pre>
<p>Indeed, there are two GPLs, GPL96 and GPL97, as their platforms
(which we could have determined by looking at the GPLList for
<code>gse</code>). We can filter the original GSMList to include only
those GSMs with the GPL96 platform and use this list for further
processing</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gsmlist</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/funprog.html" class="external-link">Filter</a></span><span class="op">(</span><span class="kw">function</span><span class="op">(</span><span class="va">gsm</span><span class="op">)</span> <span class="op">{</span><span class="fu"><a href="../reference/GEOData-accessors.html">Meta</a></span><span class="op">(</span><span class="va">gsm</span><span class="op">)</span><span class="op">$</span><span class="va">platform_id</span><span class="op">==</span><span class="st">'GPL96'</span><span class="op">}</span>,<span class="fu"><a href="../reference/GEOData-accessors.html">GSMList</a></span><span class="op">(</span><span class="va">gse</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">gsmlist</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 17</span></span></code></pre>
<p>So, now we would like to know what column represents the data that we
would like to extract. Looking at the first few rows of the Table of a
single GSM will likely give us an idea (and by the way, GEO uses a
convention that the column that contains the single measurement for each
array is called the <code>VALUE</code> column, which we could use if we
don’t know what other column is most relevant).</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/GEOData-accessors.html">Table</a></span><span class="op">(</span><span class="va">gsmlist</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>,<span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">##           ID_REF  VALUE ABS_CALL</span></span>
<span><span class="co">## 1 AFFX-BioB-5_at  953.9        P</span></span>
<span><span class="co">## 2 AFFX-BioB-M_at 2982.8        P</span></span>
<span><span class="co">## 3 AFFX-BioB-3_at 1657.9        P</span></span>
<span><span class="co">## 4 AFFX-BioC-5_at 2652.7        P</span></span>
<span><span class="co">## 5 AFFX-BioC-3_at 2019.5        P</span></span></code></pre>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># and get the column descriptions</span></span>
<span><span class="fu"><a href="../reference/GEOData-accessors.html">Columns</a></span><span class="op">(</span><span class="va">gsmlist</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>,<span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">##        Column</span></span>
<span><span class="co">## 1            </span></span>
<span><span class="co">## 2       VALUE</span></span>
<span><span class="co">## 3    ABS_CALL</span></span>
<span><span class="co">## NA       &lt;NA&gt;</span></span>
<span><span class="co">## NA.1     &lt;NA&gt;</span></span>
<span><span class="co">##                                                                     Description</span></span>
<span><span class="co">## 1                                                                      ID_REF =</span></span>
<span><span class="co">## 2                            MAS 5.0 Statistical Algorithm (mean scaled to 500)</span></span>
<span><span class="co">## 3    MAS 5.0 Absent, Marginal, Present call  with Alpha1 = 0.05, Alpha2 = 0.065</span></span>
<span><span class="co">## NA                                                                         &lt;NA&gt;</span></span>
<span><span class="co">## NA.1                                                                       &lt;NA&gt;</span></span></code></pre>
<p>We will indeed use the <code>VALUE</code> column. We then want to
make a matrix of these values like so:</p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># get the probeset ordering</span></span>
<span><span class="va">probesets</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/GEOData-accessors.html">Table</a></span><span class="op">(</span><span class="fu"><a href="../reference/GEOData-accessors.html">GPLList</a></span><span class="op">(</span><span class="va">gse</span><span class="op">)</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">$</span><span class="va">ID</span></span>
<span><span class="co"># make the data matrix from the VALUE columns from each GSM</span></span>
<span><span class="co"># being careful to match the order of the probesets in the platform</span></span>
<span><span class="co"># with those in the GSMs</span></span>
<span><span class="va">data.matrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/do.call.html" class="external-link">do.call</a></span><span class="op">(</span><span class="st">'cbind'</span>,<span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">gsmlist</span>,<span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> </span>
<span>                                      <span class="op">{</span><span class="va">tab</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/GEOData-accessors.html">Table</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span></span>
<span>                                       <span class="va">mymatch</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/match.html" class="external-link">match</a></span><span class="op">(</span><span class="va">probesets</span>,<span class="va">tab</span><span class="op">$</span><span class="va">ID_REF</span><span class="op">)</span></span>
<span>                                       <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="va">tab</span><span class="op">$</span><span class="va">VALUE</span><span class="op">[</span><span class="va">mymatch</span><span class="op">]</span><span class="op">)</span></span>
<span>                                     <span class="op">}</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">data.matrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="va">data.matrix</span>,<span class="fl">2</span>,<span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span><span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">)</span><span class="op">}</span><span class="op">)</span></span>
<span><span class="va">data.matrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log2</a></span><span class="op">(</span><span class="va">data.matrix</span><span class="op">)</span></span>
<span><span class="va">data.matrix</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>,<span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">##       GSM11805  GSM11814  GSM11823  GSM11830  GSM12067  GSM12075  GSM12079</span></span>
<span><span class="co">## [1,] 10.926963 11.105254 11.275019 11.438636 11.424376 11.222795 11.469845</span></span>
<span><span class="co">## [2,]  5.749534  7.908092  7.093814  7.514122  7.901470  6.407693  5.165912</span></span>
<span><span class="co">## [3,]  7.066089  7.750205  7.244126  7.962896  7.337176  6.569856  7.477354</span></span>
<span><span class="co">## [4,] 12.660353 12.479755 12.215897 11.458355 11.397568 12.529870 12.240046</span></span>
<span><span class="co">## [5,]  6.195741  6.061776  6.565293  6.583459  6.877744  6.652486  3.981853</span></span>
<span><span class="co">##       GSM12098  GSM12100  GSM12105  GSM12268  GSM12270  GSM12283  GSM12298</span></span>
<span><span class="co">## [1,] 10.823367 10.835971 10.810893 11.062653 10.323055 11.181028 11.566387</span></span>
<span><span class="co">## [2,]  6.556123  8.207014  6.816344  6.563768  7.353147  5.770829  6.912889</span></span>
<span><span class="co">## [3,]  7.708739  7.428779  7.754888  7.126188  8.742815  7.339850  7.602142</span></span>
<span><span class="co">## [4,] 12.336534 11.762839 11.237509 12.412490 11.213408 12.678380 12.232901</span></span>
<span><span class="co">## [5,]  5.501439  6.247928  6.017922  6.525129  6.683696  5.918863  5.837943</span></span>
<span><span class="co">##       GSM12300  GSM12399  GSM12444</span></span>
<span><span class="co">## [1,] 11.078151 11.535178 11.105450</span></span>
<span><span class="co">## [2,]  4.812498  7.471675  7.488644</span></span>
<span><span class="co">## [3,]  7.383704  7.432959  7.381110</span></span>
<span><span class="co">## [4,] 12.090939 11.421802 12.172834</span></span>
<span><span class="co">## [5,]  6.281698  5.419539  5.469235</span></span></code></pre>
<p>Note that we do a <code>match</code> to make sure that the values and
the platform information are in the same order. Finally, to make the
<code>ExpressionSet</code> object:</p>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/Biobase" class="external-link">Biobase</a></span><span class="op">)</span></span>
<span><span class="co"># go through the necessary steps to make a compliant ExpressionSet</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">data.matrix</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">probesets</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">data.matrix</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">gsmlist</span><span class="op">)</span></span>
<span><span class="va">pdata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>samples<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">gsmlist</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">pdata</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">gsmlist</span><span class="op">)</span></span>
<span><span class="va">pheno</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/methods/as.html" class="external-link">as</a></span><span class="op">(</span><span class="va">pdata</span>,<span class="st">"AnnotatedDataFrame"</span><span class="op">)</span></span>
<span><span class="va">eset2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/methods/new.html" class="external-link">new</a></span><span class="op">(</span><span class="st">'ExpressionSet'</span>,exprs<span class="op">=</span><span class="va">data.matrix</span>,phenoData<span class="op">=</span><span class="va">pheno</span><span class="op">)</span></span>
<span><span class="va">eset2</span></span></code></pre></div>
<pre><code><span><span class="co">## ExpressionSet (storageMode: lockedEnvironment)</span></span>
<span><span class="co">## assayData: 22283 features, 17 samples </span></span>
<span><span class="co">##   element names: exprs </span></span>
<span><span class="co">## protocolData: none</span></span>
<span><span class="co">## phenoData</span></span>
<span><span class="co">##   sampleNames: GSM11805 GSM11814 ... GSM12444 (17 total)</span></span>
<span><span class="co">##   varLabels: samples</span></span>
<span><span class="co">##   varMetadata: labelDescription</span></span>
<span><span class="co">## featureData: none</span></span>
<span><span class="co">## experimentData: use 'experimentData(object)'</span></span>
<span><span class="co">## Annotation:</span></span></code></pre>
<p>So, using a combination of <code>lapply</code> on the GSMList, one
can extract as many columns of interest as necessary to build the data
structure of choice. Because the GSM data from the GEO website are fully
downloaded and included in the <code>GSE</code> object, one can extract
foreground and background as well as quality for two-channel arrays, for
example. Getting array annotation is also a bit more complicated, but by
replacing ``platform’’ in the lapply call to get platform information
for each array, one can get other information associated with each
array.</p>
</div>
</div>
<div class="section level2">
<h2 id="accessing-raw-data-from-geo">Accessing Raw Data from GEO<a class="anchor" aria-label="anchor" href="#accessing-raw-data-from-geo"></a>
</h2>
<p>NCBI GEO accepts (but has not always required) raw data such as .CEL
files, .CDF files, images, etc. Sometimes, it is useful to get quick
access to such data. A single function, <code>getGEOSuppFiles</code>,
can take as an argument a GEO accession and will download all the raw
data associate with that accession. By default, the function will create
a directory in the current working directory to store the raw data for
the chosen GEO accession. Combining a simple <code>sapply</code>
statement or other loop structure with <code>getGEOSuppFiles</code>
makes for a very simple way to get gobs of raw data quickly and easily
without needing to know the specifics of GEO raw data URLs.</p>
</div>
<div class="section level2">
<h2 id="use-cases">Use Cases<a class="anchor" aria-label="anchor" href="#use-cases"></a>
</h2>
<p>GEOquery can be quite powerful for gathering a lot of data quickly. A
few examples can be useful to show how this might be done for data
mining purposes.</p>
<div class="section level3">
<h3 id="getting-all-series-records-for-a-given-platform">Getting all Series Records for a Given Platform<a class="anchor" aria-label="anchor" href="#getting-all-series-records-for-a-given-platform"></a>
</h3>
<p>For data mining purposes, it is sometimes useful to be able to pull
all the GSE records for a given platform. GEOquery makes this very easy,
but a little bit of knowledge of the GPL record is necessary to get
started. The GPL record contains both the GSE and GSM accessions that
reference it. Some code is useful to illustrate the point:</p>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gpl97</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getGEO.html">getGEO</a></span><span class="op">(</span><span class="st">'GPL97'</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/GEOData-accessors.html">Meta</a></span><span class="op">(</span><span class="va">gpl97</span><span class="op">)</span><span class="op">$</span><span class="va">title</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "[HG-U133B] Affymetrix Human Genome U133B Array"</span></span></code></pre>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="../reference/GEOData-accessors.html">Meta</a></span><span class="op">(</span><span class="va">gpl97</span><span class="op">)</span><span class="op">$</span><span class="va">series_id</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "GSE362" "GSE473" "GSE620" "GSE674" "GSE781" "GSE907"</span></span></code></pre>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="../reference/GEOData-accessors.html">Meta</a></span><span class="op">(</span><span class="va">gpl97</span><span class="op">)</span><span class="op">$</span><span class="va">series_id</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 165</span></span></code></pre>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="../reference/GEOData-accessors.html">Meta</a></span><span class="op">(</span><span class="va">gpl97</span><span class="op">)</span><span class="op">$</span><span class="va">sample_id</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "GSM3922" "GSM3924" "GSM3926" "GSM3928" "GSM3930" "GSM3932"</span></span></code></pre>
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="../reference/GEOData-accessors.html">Meta</a></span><span class="op">(</span><span class="va">gpl97</span><span class="op">)</span><span class="op">$</span><span class="va">sample_id</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 7917</span></span></code></pre>
<p>The code above loads the GPL97 record into R. The Meta method
extracts a list of header information from the GPL record. The
<code>title</code> gives the human name of the platform. The
<code>series_id</code> gives a vector of series ids. Note that there are
165 series associated with this platform and 7917 samples. Code like the
following could be used to download all the samples or series. I show
only the first 5 samples as an example:</p>
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gsmids</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/GEOData-accessors.html">Meta</a></span><span class="op">(</span><span class="va">gpl97</span><span class="op">)</span><span class="op">$</span><span class="va">sample_id</span></span>
<span><span class="va">gsmlist</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">gsmids</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span>,<span class="va">getGEO</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">gsmlist</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "GSM3922" "GSM3924" "GSM3926" "GSM3928" "GSM3930"</span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="conclusion">Conclusion<a class="anchor" aria-label="anchor" href="#conclusion"></a>
</h2>
<p>The GEOquery package provides a bridge to the vast array resources
contained in the NCBI GEO repositories. By maintaining the full richness
of the GEO data rather than focusing on getting only the ``numbers’’, it
is possible to integrate GEO data into current Bioconductor data
structures and to perform analyses on that data quite quickly and
easily. These tools will hopefully open GEO data more fully to the array
community at large.</p>
<div class="section level3">
<h3 id="citing-geoquery">Citing GEOquery<a class="anchor" aria-label="anchor" href="#citing-geoquery"></a>
</h3>
<p>Please consider citing GEOquery if used in support of your own
research:</p>
<div class="sourceCode" id="cb58"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/citation.html" class="external-link">citation</a></span><span class="op">(</span><span class="st">"GEOquery"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Please cite the following if utilizing the GEOquery software:</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##   Davis S, Meltzer P (2007). "GEOquery: a bridge between the Gene</span></span>
<span><span class="co">##   Expression Omnibus (GEO) and BioConductor." _Bioinformatics_, *14*,</span></span>
<span><span class="co">##   1846-1847. doi:10.1093/bioinformatics/btm254</span></span>
<span><span class="co">##   &lt;https://doi.org/10.1093/bioinformatics/btm254&gt;.</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## A BibTeX entry for LaTeX users is</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##   @Article{,</span></span>
<span><span class="co">##     author = {Sean Davis and Paul Meltzer},</span></span>
<span><span class="co">##     title = {GEOquery: a bridge between the Gene Expression Omnibus (GEO) and BioConductor},</span></span>
<span><span class="co">##     journal = {Bioinformatics},</span></span>
<span><span class="co">##     year = {2007},</span></span>
<span><span class="co">##     volume = {14},</span></span>
<span><span class="co">##     pages = {1846--1847},</span></span>
<span><span class="co">##     doi = {10.1093/bioinformatics/btm254},</span></span>
<span><span class="co">##   }</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="reporting-problems-or-bugs">Reporting problems or bugs<a class="anchor" aria-label="anchor" href="#reporting-problems-or-bugs"></a>
</h3>
<p>If you run into problems using GEOquery, the <a href="https://support.bioconductor.org/" class="external-link">Bioconductor Support site</a>
is a good first place to ask for help. If you are convinced that there
is a bug in GEOquery (this is pretty unusual, but not unheard of), feel
free to submit an issue on the <a href="https://github.com/seandavi/GEOquery" class="external-link">GEOquery github site</a> or
file a bug report directly from R (will open a new github issue):</p>
<div class="sourceCode" id="cb60"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/bug.report.html" class="external-link">bug.report</a></span><span class="op">(</span>package<span class="op">=</span><span class="st">'GEOquery'</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<p>The following package and versions were used in the production of
this vignette.</p>
<pre><code><span><span class="co">## R version 4.4.1 (2024-06-14)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">## Running under: Ubuntu 22.04.4 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">##  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">## [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: UTC</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">## [1] GEOquery_2.73.4     Biobase_2.64.0      BiocGenerics_0.50.0</span></span>
<span><span class="co">## [4] knitr_1.48         </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##  [1] SummarizedExperiment_1.34.0 httr2_1.0.2                </span></span>
<span><span class="co">##  [3] xfun_0.46                   bslib_0.8.0                </span></span>
<span><span class="co">##  [5] lattice_0.22-6              tzdb_0.4.0                 </span></span>
<span><span class="co">##  [7] vctrs_0.6.5                 tools_4.4.1                </span></span>
<span><span class="co">##  [9] generics_0.1.3              curl_5.2.1                 </span></span>
<span><span class="co">## [11] stats4_4.4.1                tibble_3.2.1               </span></span>
<span><span class="co">## [13] fansi_1.0.6                 R.oo_1.26.0                </span></span>
<span><span class="co">## [15] pkgconfig_2.0.3             Matrix_1.7-0               </span></span>
<span><span class="co">## [17] data.table_1.15.4           rentrez_1.2.3              </span></span>
<span><span class="co">## [19] desc_1.4.3                  S4Vectors_0.42.1           </span></span>
<span><span class="co">## [21] lifecycle_1.0.4             GenomeInfoDbData_1.2.12    </span></span>
<span><span class="co">## [23] compiler_4.4.1              textshaping_0.4.0          </span></span>
<span><span class="co">## [25] statmod_1.5.0               GenomeInfoDb_1.40.1        </span></span>
<span><span class="co">## [27] htmltools_0.5.8.1           sass_0.4.9                 </span></span>
<span><span class="co">## [29] yaml_2.3.10                 pkgdown_2.1.0              </span></span>
<span><span class="co">## [31] pillar_1.9.0                crayon_1.5.3               </span></span>
<span><span class="co">## [33] jquerylib_0.1.4             tidyr_1.3.1                </span></span>
<span><span class="co">## [35] R.utils_2.12.3              limma_3.60.4               </span></span>
<span><span class="co">## [37] DelayedArray_0.30.1         cachem_1.1.0               </span></span>
<span><span class="co">## [39] abind_1.4-5                 tidyselect_1.2.1           </span></span>
<span><span class="co">## [41] digest_0.6.36               dplyr_1.1.4                </span></span>
<span><span class="co">## [43] purrr_1.0.2                 fastmap_1.2.0              </span></span>
<span><span class="co">## [45] grid_4.4.1                  cli_3.6.3                  </span></span>
<span><span class="co">## [47] SparseArray_1.4.8           magrittr_2.0.3             </span></span>
<span><span class="co">## [49] S4Arrays_1.4.1              XML_3.99-0.17              </span></span>
<span><span class="co">## [51] utf8_1.2.4                  readr_2.1.5                </span></span>
<span><span class="co">## [53] rappdirs_0.3.3              UCSC.utils_1.0.0           </span></span>
<span><span class="co">## [55] rmarkdown_2.27              XVector_0.44.0             </span></span>
<span><span class="co">## [57] httr_1.4.7                  matrixStats_1.3.0          </span></span>
<span><span class="co">## [59] R.methodsS3_1.8.2           ragg_1.3.2                 </span></span>
<span><span class="co">## [61] hms_1.1.3                   evaluate_0.24.0            </span></span>
<span><span class="co">## [63] GenomicRanges_1.56.1        IRanges_2.38.1             </span></span>
<span><span class="co">## [65] rlang_1.1.4                 glue_1.7.0                 </span></span>
<span><span class="co">## [67] xml2_1.3.6                  jsonlite_1.8.8             </span></span>
<span><span class="co">## [69] R6_2.5.1                    MatrixGenerics_1.16.0      </span></span>
<span><span class="co">## [71] systemfonts_1.1.0           fs_1.6.4                   </span></span>
<span><span class="co">## [73] zlibbioc_1.50.0</span></span></code></pre>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Sean Davis.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

    </footer>
</div>





  </body>
</html>
